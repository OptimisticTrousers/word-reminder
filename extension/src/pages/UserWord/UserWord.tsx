import {
  Definition,
  Detail,
  Meaning,
  Phonetic,
  User,
  UserWord as IUserWord,
  Word,
  Image,
} from "common";
import { Play, Trash } from "lucide-react";
import CSSModules from "react-css-modules";
import { Link, useOutletContext, useParams } from "react-router-dom";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { Loading } from "../../components/ui/Loading";
import { Error500 } from "../Error500";
import { userWordService } from "../../services/user_word_service";
import styles from "./UserWord.module.css";
import { ImageCarousel } from "../../components/ui/ImageCarousel";
import { FormEvent, useState } from "react";
import { DeleteUserWordModal } from "../../components/modals/DeleteUserWordModal";
import { useNotificationError } from "../../hooks/useNotificationError";

export const UserWord = CSSModules(
  function () {
    const { user }: { user: User } = useOutletContext();
    const { userWordId } = useParams();
    const userId = String(user.id);
    const queryClient = useQueryClient();

    const { data, isLoading, failureReason } = useQuery({
      queryKey: ["userWords", userWordId],
      queryFn: () => {
        return userWordService.getUserWord({
          userId,
          userWordId: String(userWordId),
        });
      },
    });
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
    const { showNotificationError } = useNotificationError();

    const { isPending, mutate } = useMutation({
      mutationFn: userWordService.updateUserWord,
      onSuccess: () => {
        queryClient.invalidateQueries({
          queryKey: ["userWords"],
        });
      },
      onError: showNotificationError,
    });

    function handleLearned(event: FormEvent<HTMLFormElement>) {
      event.preventDefault();
      const formData = new FormData(event.currentTarget);
      mutate({
        userId,
        userWordId: String(userWordId),
        body: {
          learned: (formData.get("learned") as string) === "true",
        },
      });
    }

    const disabled = isPending;

    if (failureReason) {
      return <Error500 message={failureReason.message} />;
    }

    if (isLoading) {
      return <Loading />;
    }

    function handleAudio(url: string) {
      const audio = new Audio(url);
      audio.play();
    }

    function toggleDeleteModal() {
      setIsDeleteModalOpen((prevIsDeleteModalOpen) => !prevIsDeleteModalOpen);
    }

    const json = data?.json;
    const userWord: (IUserWord & { word: Word; images: Image[] }) | undefined =
      json?.userWord;
    const details = userWord?.word.details;
    const images = userWord?.images ?? [];

    return (
      details && (
        <div styleName="word">
          {isDeleteModalOpen && (
            <DeleteUserWordModal
              toggleModal={toggleDeleteModal}
              userWordId={String(userWord.id)}
            />
          )}
          {<h2 styleName="word__heading">{details[0].word}</h2>}
          {
            <p styleName="word__learned">
              {userWord.learned ? "Learned word" : "Word not learned"}
            </p>
          }
          {
            <p styleName="word__created_at">
              {userWord.created_at.toLocaleString()}
            </p>
          }
          {<ImageCarousel images={images} />}
          <div styleName="word__buttons">
            <form styleName="word__form" onSubmit={handleLearned}>
              <input
                styleName="word__input word__input--hidden"
                type="hidden"
                id="learned"
                name="learned"
                required
                defaultValue={String(!userWord.learned)}
              />
              <button
                styleName="word__button word__button--learned"
                type="submit"
                disabled={disabled}
              >
                {disabled ? "Toggling Learned..." : "Toggle Learned"}
              </button>
            </form>
            <button
              styleName="word__delete word__button--delete"
              onClick={toggleDeleteModal}
              aria-haspopup={true}
              aria-label="Open delete user word modal"
            >
              <Trash styleName="word__icon" />
            </button>
          </div>
          {details.map((detail: Detail, index: number) => {
            return (
              <div key={index} styleName="word_content">
                <div styleName="word__top">
                  <div styleName="word__title">
                    {detail.phonetic && (
                      <p styleName="word__phonetic">{detail.phonetic}</p>
                    )}
                    <ul styleName="word__phonetics">
                      {detail.phonetics &&
                        detail.phonetics.map(
                          (phonetic: Phonetic, index: number) => {
                            return (
                              <li key={index} styleName="word__phonetic">
                                {phonetic.text && (
                                  <p styleName="word__text">{phonetic.text}</p>
                                )}
                                {phonetic.audio && (
                                  <button
                                    styleName="word__button"
                                    onClick={() => handleAudio(phonetic.audio!)}
                                  >
                                    <span styleName="word__text">
                                      Pronounce word
                                    </span>
                                    <Play styleName="word__icon" />
                                  </button>
                                )}
                                {phonetic.sourceUrl && (
                                  <Link
                                    to={phonetic.sourceUrl}
                                    styleName="word__link"
                                  >
                                    Phonetic Source
                                  </Link>
                                )}
                                {phonetic.license && (
                                  <div styleName="word__license">
                                    <p styleName="word__name">
                                      {phonetic.license.name}
                                    </p>
                                    <Link
                                      to={phonetic.license.url}
                                      styleName="word__url"
                                    >
                                      Phonetic License
                                    </Link>
                                  </div>
                                )}
                              </li>
                            );
                          }
                        )}
                    </ul>
                  </div>
                </div>
                <div styleName="word__container">
                  <div styleName="word__block">
                    <h3 styleName="word__subtitle">Origin</h3>
                    <p styleName="word__description">{detail.origin}</p>
                  </div>
                  {detail.meanings.map((meaning: Meaning, index: number) => (
                    <div styleName="word__meanings" key={index}>
                      <span styleName="word__part-of-speech">
                        {meaning.partOfSpeech} <hr styleName="word__line" />
                      </span>
                      <div styleName="word__block">
                        <h3 styleName="word__subtitle">Meanings</h3>
                        <ul styleName="word__definitions">
                          {meaning.definitions.map(
                            (definition: Definition, index: number) => (
                              <li styleName="word__definition" key={index}>
                                <p styleName="word__definition">
                                  {definition.definition}
                                </p>
                                {definition.example && (
                                  <p styleName="word__example">
                                    {definition.example}
                                  </p>
                                )}
                                <ul styleName="word__synonyms">
                                  {definition.synonyms &&
                                    definition.synonyms.map(
                                      (synonym: string, index: number) => {
                                        return (
                                          <li
                                            styleName="word__synonym"
                                            key={index}
                                          >
                                            {synonym}
                                          </li>
                                        );
                                      }
                                    )}
                                </ul>
                                <ul styleName="word__antonyms">
                                  {definition.antonyms &&
                                    definition.antonyms.map(
                                      (antonym: string, index: number) => {
                                        return (
                                          <li
                                            styleName="word__antonym"
                                            key={index}
                                          >
                                            {antonym}
                                          </li>
                                        );
                                      }
                                    )}
                                </ul>
                              </li>
                            )
                          )}
                        </ul>
                      </div>
                      {meaning.synonyms && (
                        <div styleName="word__block">
                          <h3 styleName="word__subtitle">Synonyms</h3>
                          <ul styleName="word__definitions">
                            {meaning.synonyms.map(
                              (synonym: string, index: number) => {
                                return (
                                  <li styleName="word__synonym" key={index}>
                                    {synonym}
                                  </li>
                                );
                              }
                            )}
                          </ul>
                        </div>
                      )}
                      {meaning.antonyms && (
                        <div styleName="word__block">
                          <h3 styleName="word__subtitle">Antonyms</h3>
                          <ul styleName="word__definitions">
                            {meaning.antonyms.map(
                              (antonym: string, index: number) => {
                                return (
                                  <li styleName="word__antonym" key={index}>
                                    {antonym}
                                  </li>
                                );
                              }
                            )}
                          </ul>
                        </div>
                      )}
                      {detail.license && (
                        <div styleName="word__license">
                          <p styleName="word__name">{detail.license.name}</p>
                          <Link to={detail.license.url} styleName="word__url">
                            Word License
                          </Link>
                        </div>
                      )}
                      <ul styleName="word__list">
                        {detail.sourceUrls &&
                          detail.sourceUrls.map(
                            (sourceUrl: string, index: number) => {
                              return (
                                <li key={index} styleName="word__item">
                                  <Link to={sourceUrl} styleName="word__link">
                                    Word Source
                                  </Link>
                                </li>
                              );
                            }
                          )}
                      </ul>
                    </div>
                  ))}
                </div>
              </div>
            );
          })}
        </div>
      )
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
