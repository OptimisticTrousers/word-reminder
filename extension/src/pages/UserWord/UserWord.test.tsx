import { render, screen } from "@testing-library/react";
import { wordService } from "../../services/word_service";
import { UserWord } from "./UserWord";
import userEvent from "@testing-library/user-event";
import { createRoutesStub, Outlet } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

vi.mock("../../components/ui/Loading", () => {
  return {
    Loading: function () {
      return <div data-testid="loading">Loading...</div>;
    },
  };
});

vi.mock("../Error500", () => {
  return {
    Error500: function ({ message }: { message: string }) {
      return <div data-testid="error">{message}</div>;
    },
  };
});

describe("UserWord component", () => {
  const testUser = {
    id: "1",
  };
  const wordId = "1";
  const status = 200;
  const created_at = new Date("2020-05-12T23:50:21.817Z");

  function setup() {
    const queryClient = new QueryClient();

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "words/:wordId",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <UserWord />
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={[`/words/${wordId}`]} />),
    };
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders word, origin, phonetics, meanings, license, source urls, and images", async () => {
    const json = {
      userWord: {
        created_at,
        learned: true,
        images: [{ src: "/image1" }, { src: "/image2" }],
        details: [
          {
            word: "word",
            phonetic: "/wɜːd/",
            origin: "The foundation of a language.",
            phonetics: [
              {
                text: "/wɜːd/",
                audio: "",
              },
              {
                text: "/wɝd/",
                audio:
                  "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3",
                sourceUrl:
                  "https://commons.wikimedia.org/w/index.php?curid=1197728",
                license: {
                  name: "BY-SA 3.0",
                  url: "https://creativecommons.org/licenses/by-sa/3.0",
                },
              },
            ],
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
            license: {
              name: "CC BY-SA 3.0",
              url: "https://creativecommons.org/licenses/by-sa/3.0",
            },
            sourceUrls: ["https://en.wiktionary.org/wiki/word"],
          },
        ],
      },
    };
    const mockGetUserWord = vi
      .spyOn(wordService, "getUserWord")
      .mockImplementation(async () => {
        return { json, status };
      });

    const { asFragment } = setup();

    const createdAt = await screen.findByText(
      json.userWord.created_at.toLocaleString()
    );
    const word = screen.getByRole("heading", { name: "word" });
    const phonetic = screen.getByText("/wɝd/");
    const origin = screen.getByText("The foundation of a language.");
    const phoneticSourceURl = screen.getByRole("link", {
      name: "Phonetic Source",
    });
    const phoneticLicenseName = screen.getByText("BY-SA 3.0");
    const phoneticLicenseUrl = screen.getByRole("link", {
      name: "Phonetic License",
    });
    const partOfSpeech = screen.getByText("noun");
    const definition1 = screen.getByText(
      "News; tidings (used without an article)."
    );
    const example1 = screen.getByText("Have you had any word from John yet?");
    const definition2 = screen.getByText("A promise; an oath or guarantee.");
    const example2 = screen.getByText(
      "I give you my word that I will be there on time."
    );
    const synonym1 = screen.getByText("promise");
    const synonym2 = screen.getByText("Bible");
    const synonym3 = screen.getByText("Logos");
    const synonym4 = screen.getByText("vocable");
    const antonym1 = screen.getByText("false promise");
    const wordLicenseName = screen.getByText("CC BY-SA 3.0");
    const wordLicenseUrl = screen.getByRole("link", {
      name: "Word License",
    });
    const wordSourceUrl = screen.getByRole("link", { name: "Word Source" });
    const [image1, image2] = screen.getAllByRole("presentation");
    const pronounciationButton = screen.getByRole("button", {
      name: "Pronounce word",
    });
    const loading = screen.queryByTestId("loading");
    const error = screen.queryByTestId("error");
    expect(createdAt).toBeInTheDocument();
    expect(word).toBeInTheDocument();
    expect(phonetic).toBeInTheDocument();
    expect(origin).toBeInTheDocument();
    expect(phoneticSourceURl).toBeInTheDocument();
    expect(phoneticLicenseName).toBeInTheDocument();
    expect(phoneticLicenseUrl).toBeInTheDocument();
    expect(partOfSpeech).toBeInTheDocument();
    expect(definition1).toBeInTheDocument();
    expect(example1).toBeInTheDocument();
    expect(definition2).toBeInTheDocument();
    expect(example2).toBeInTheDocument();
    expect(synonym1).toBeInTheDocument();
    expect(synonym2).toBeInTheDocument();
    expect(synonym3).toBeInTheDocument();
    expect(synonym4).toBeInTheDocument();
    expect(antonym1).toBeInTheDocument();
    expect(wordLicenseName).toBeInTheDocument();
    expect(wordLicenseUrl).toBeInTheDocument();
    expect(wordSourceUrl).toBeInTheDocument();
    expect(image1.getAttribute("src")).toBe("/image1");
    expect(image2.getAttribute("src")).toBe("/image2");
    expect(image1.getAttribute("alt")).toBe("");
    expect(image2.getAttribute("alt")).toBe("");
    expect(pronounciationButton).toBeInTheDocument();
    expect(loading).not.toBeInTheDocument();
    expect(error).not.toBeInTheDocument();
    expect(mockGetUserWord).toHaveBeenCalledTimes(1);
    expect(mockGetUserWord).toHaveBeenCalledWith(testUser.id, wordId);
    expect(asFragment()).toMatchSnapshot();
  });

  describe("learned", () => {
    it("tells the user that they already learned the word when learned is true", async () => {
      const json = {
        userWord: {
          created_at,
          learned: true,
          details: [
            {
              word: "word",
              meanings: [
                {
                  partOfSpeech: "noun",
                  definitions: [
                    {
                      definition: "News; tidings (used without an article).",
                      synonyms: [],
                      antonyms: [],
                      example: "Have you had any word from John yet?",
                    },
                    {
                      definition: "A promise; an oath or guarantee.",
                      synonyms: ["promise"],
                      antonyms: ["false promise"],
                      example:
                        "I give you my word that I will be there on time.",
                    },
                  ],
                  synonyms: ["Bible", "Logos", "vocable"],
                  antonyms: [],
                },
              ],
            },
          ],
          images: [],
        },
      };
      vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
        return { json, status };
      });

      setup();

      const learned = await screen.findByText("Learned word");
      expect(learned).toBeInTheDocument();
    });

    it("tells the user that they have not learned the word when learned is false", async () => {
      const json = {
        userWord: {
          created_at,
          learned: false,
          details: [
            {
              word: "word",
              meanings: [
                {
                  partOfSpeech: "noun",
                  definitions: [
                    {
                      definition: "News; tidings (used without an article).",
                      synonyms: [],
                      antonyms: [],
                      example: "Have you had any word from John yet?",
                    },
                    {
                      definition: "A promise; an oath or guarantee.",
                      synonyms: ["promise"],
                      antonyms: ["false promise"],
                      example:
                        "I give you my word that I will be there on time.",
                    },
                  ],
                  synonyms: ["Bible", "Logos", "vocable"],
                  antonyms: [],
                },
              ],
            },
          ],
          images: [],
        },
      };
      vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
        return { json, status };
      });
      setup();

      const notLearned = await screen.findByText("Word not learned");
      expect(notLearned).toBeInTheDocument();
    });
  });

  describe("phonetics", () => {
    it("renders a phonetic when phonetics is available but phonetic is not available", async () => {
      const json = {
        userWord: {
          created_at,
          learned: false,
          images: [],
          details: [
            {
              id: "1",
              word: "word",
              phonetics: [
                {
                  text: "/wɜːd/",
                  audio: "",
                },
                {
                  text: "/wɝd/",
                  audio:
                    "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3",
                  sourceUrl:
                    "https://commons.wikimedia.org/w/index.php?curid=1197728",
                  license: {
                    name: "BY-SA 3.0",
                    url: "https://creativecommons.org/licenses/by-sa/3.0",
                  },
                },
              ],
              meanings: [
                {
                  partOfSpeech: "noun",
                  definitions: [
                    {
                      definition: "News; tidings (used without an article).",
                      synonyms: [],
                      antonyms: [""],
                      example: "Have you had any word from John yet?",
                    },
                    {
                      definition: "A promise; an oath or guarantee.",
                      synonyms: ["promise"],
                      antonyms: ["false promise"],
                      example:
                        "I give you my word that I will be there on time.",
                    },
                  ],
                  synonyms: ["Bible", "Logos", "vocable"],
                  antonyms: [],
                },
              ],
            },
          ],
        },
      };
      vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
        return { json, status };
      });

      setup();

      const phonetic = screen.queryByText("/wɜːd/");

      expect(phonetic).not.toBeInTheDocument();
    });

    it("renders a phonetic when phonetic is available but phonetic is not available", async () => {
      const json = {
        created_at,
        learned: false,
        images: [],
        userWord: [
          {
            id: "1",
            word: "word",
            phonetic: "/wɜːd/",
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [""],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
          },
        ],
      };
      vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
        return { json, status };
      });

      setup();

      const phonetic = screen.queryByText("/wɜːd/");

      expect(phonetic).not.toBeInTheDocument();
    });

    it("does not render a phonetic when not available", async () => {
      const json = {
        userWord: {
          created_at,
          learned: false,
          images: [],
          details: [
            {
              id: "1",
              word: "word",
              meanings: [
                {
                  partOfSpeech: "noun",
                  definitions: [
                    {
                      definition: "News; tidings (used without an article).",
                      synonyms: [],
                      antonyms: [""],
                      example: "Have you had any word from John yet?",
                    },
                    {
                      definition: "A promise; an oath or guarantee.",
                      synonyms: ["promise"],
                      antonyms: ["false promise"],
                      example:
                        "I give you my word that I will be there on time.",
                    },
                  ],
                  synonyms: ["Bible", "Logos", "vocable"],
                  antonyms: [],
                },
              ],
            },
          ],
        },
      };
      vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
        return { json, status };
      });

      setup();

      const phonetic = screen.queryByText("/wɜːd/");

      expect(phonetic).not.toBeInTheDocument();
    });
  });

  it("does not render origin when not available", async () => {
    const json = {
      userWord: {
        created_at,
        learned: false,
        images: [],
        details: [
          {
            id: "1",
            word: "word",
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [""],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
          },
        ],
      },
    };
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return { json, status };
    });

    setup();

    const origin = screen.queryByText("The foundation of a language.");
    expect(origin).not.toBeInTheDocument();
  });

  it("does not render audio pronounciation when not available", () => {
    const json = {
      userWord: {
        created_at,
        learned: false,
        images: [],
        details: [
          {
            id: "1",
            word: "word",
            phonetics: [
              {
                text: "/wɜːd/",
                audio: "",
              },
              {
                text: "/wɝd/",
                audio: "",
                sourceUrl:
                  "https://commons.wikimedia.org/w/index.php?curid=1197728",
                license: {
                  name: "BY-SA 3.0",
                  url: "https://creativecommons.org/licenses/by-sa/3.0",
                },
              },
            ],
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [""],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
          },
        ],
      },
    };
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return { json, status };
    });

    setup();

    const pronounciationButton = screen.queryByRole("button", {
      name: "Pronounce word",
    });
    expect(pronounciationButton).not.toBeInTheDocument();
  });

  it("does not render licenses when not available", async () => {
    const json = {
      userWord: {
        created_at,
        learned: false,
        images: [],
        details: [
          {
            id: "1",
            word: "word",
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [""],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
            sourceUrls: ["https://en.wiktionary.org/wiki/word"],
          },
        ],
      },
    };
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return { json, status };
    });

    setup();

    const phoneticLicenseName = screen.queryByText("BY-SA 3.0");
    const phoneticLicenseUrl = screen.queryByRole("link", {
      name: "Phonetic License URL",
    });
    const wordLicenseName = screen.queryByText("CC BY-SA 3.0");
    const wordLicenseUrl = screen.queryByRole("link", {
      name: "Word License URL",
    });

    expect(phoneticLicenseName).not.toBeInTheDocument();
    expect(phoneticLicenseUrl).not.toBeInTheDocument();
    expect(wordLicenseName).not.toBeInTheDocument();
    expect(wordLicenseUrl).not.toBeInTheDocument();
  });

  it("does not render source urls when not available", async () => {
    const json = {
      userWord: {
        created_at,
        learned: false,
        images: [],
        details: [
          {
            id: "1",
            word: "word",
            origin: "The foundation of a language.",
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [""],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
            license: {
              name: "CC BY-SA 3.0",
              url: "https://creativecommons.org/licenses/by-sa/3.0",
            },
          },
        ],
      },
    };
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return { json, status };
    });

    setup();

    const phoneticSourceURl = screen.queryByRole("link", {
      name: "Phonetic Source URL",
    });
    const wordSourceUrl = screen.queryByRole("link", {
      name: "Word Source URL",
    });

    expect(phoneticSourceURl).not.toBeInTheDocument();
    expect(wordSourceUrl).not.toBeInTheDocument();
  });

  it("does not render images are not available", async () => {
    const json = {
      userWord: {
        created_at,
        learned: false,
        images: [],
        details: [
          {
            id: "1",
            word: "word",
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [""],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
          },
        ],
      },
    };
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return { json, status };
    });

    setup();

    const images = screen.queryAllByRole("img");
    expect(images).toHaveLength(0);
  });

  it("plays audio to pronounce word when available", async () => {
    const json = {
      userWord: {
        created_at,
        learned: false,
        images: [],
        details: [
          {
            id: "1",
            word: "word",
            phonetic: "/wɜːd/",
            phonetics: [
              {
                text: "/wɜːd/",
                audio: "",
              },
              {
                text: "/wɝd/",
                audio:
                  "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3",
                sourceUrl:
                  "https://commons.wikimedia.org/w/index.php?curid=1197728",
                license: {
                  name: "BY-SA 3.0",
                  url: "https://creativecommons.org/licenses/by-sa/3.0",
                },
              },
            ],
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [""],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
          },
        ],
      },
    };
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return { json, status };
    });
    const mockPlay = vi.fn();
    const mockCtor = vi.fn();
    globalThis.Audio = mockCtor.mockImplementation(() => {
      return {
        play: mockPlay,
      };
    });
    const user = userEvent.setup();

    setup();
    const audioButton = await screen.findByRole("button", {
      name: "Pronounce word",
    });
    await user.click(audioButton);

    expect(mockCtor).toHaveBeenCalledTimes(1);
    expect(mockCtor).toHaveBeenCalledWith(
      "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3"
    );
    expect(mockPlay).toHaveBeenCalledTimes(1);
    expect(mockPlay).toHaveBeenCalledWith();
  });

  it("returns loading component when loading", async () => {
    const json = {
      userWord: {
        created_at,
        learned: false,
        images: [],
        details: [
          {
            id: "1",
            word: "word",
            phonetic: "/wɜːd/",
            origin: "The foundation of a language.",
            meanings: [
              {
                partOfSpeech: "noun",
                definitions: [
                  {
                    definition: "News; tidings (used without an article).",
                    synonyms: [],
                    antonyms: [],
                    example: "Have you had any word from John yet?",
                  },
                  {
                    definition: "A promise; an oath or guarantee.",
                    synonyms: ["promise"],
                    antonyms: ["false promise"],
                    example: "I give you my word that I will be there on time.",
                  },
                ],
                synonyms: ["Bible", "Logos", "vocable"],
                antonyms: [],
              },
            ],
          },
        ],
      },
    };
    const delay = 500;
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return new Promise((resolve) => {
        setTimeout(() => {
          resolve({ json, status });
        }, delay);
      });
    });

    setup();

    const loading = screen.getByTestId("loading");
    expect(loading).toBeInTheDocument();
  });

  it("returns error 500 page when there is an error", async () => {
    const message = "Server Error.";
    vi.spyOn(wordService, "getUserWord").mockImplementation(async () => {
      return Promise.reject({ message, status: 500 });
    });

    setup();

    const error = await screen.findByTestId("error");
    expect(error).toHaveTextContent(message);
  });
});
