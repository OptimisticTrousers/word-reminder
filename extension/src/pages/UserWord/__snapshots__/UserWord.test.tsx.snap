// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserWord component > renders word, origin, phonetics, meanings, license, source urls, and images 1`] = `
<DocumentFragment>
  <div
    class="_word_063118"
  >
    <h2
      class="_word__heading_063118"
    >
      word
    </h2>
    <p
      class="_word__learned_063118"
    >
      Learned word
    </p>
    <p
      class="_word__created_at_063118"
    >
      5/12/2020, 7:50:21 PM
    </p>
    <div
      data-testid="image-carousel"
    >
      [object Object]
    </div>
    <div
      class="_word_content_063118"
    >
      <div
        class="_word__top_063118"
      >
        <div
          class="_word__title_063118"
        >
          <p
            class="_word__phonetic_063118"
          >
            /wɜːd/
          </p>
          <ul
            class="_word__phonetics_063118"
          >
            <li
              class="_word__phonetic_063118"
            >
              <p
                class="_word__text_063118"
              >
                /wɜːd/
              </p>
            </li>
            <li
              class="_word__phonetic_063118"
            >
              <p
                class="_word__text_063118"
              >
                /wɝd/
              </p>
              <button
                class="_word__button_063118"
              >
                <span
                  class="_word__text_063118"
                >
                  Pronounce word
                </span>
                <svg
                  class="lucide lucide-play _word__icon_063118"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="6 3 20 12 6 21 6 3"
                  />
                </svg>
              </button>
              <a
                class="_word__link_063118"
                href="https://commons.wikimedia.org/w/index.php?curid=1197728"
              >
                Phonetic Source
              </a>
              <div
                class="_word__license_063118"
              >
                <p
                  class="_word__name_063118"
                >
                  BY-SA 3.0
                </p>
                <a
                  class="_word__url_063118"
                  href="https://creativecommons.org/licenses/by-sa/3.0"
                >
                  Phonetic License
                </a>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="_word__container_063118"
      >
        <div
          class="_word__block_063118"
        >
          <h3
            class="_word__subtitle_063118"
          >
            Origin
          </h3>
          <p
            class="_word__description_063118"
          >
            The foundation of a language.
          </p>
        </div>
        <div
          class="_word__meanings_063118"
        >
          <span
            class="_word__part-of-speech_063118"
          >
            noun 
            <hr
              class="_word__line_063118"
            />
          </span>
          <div
            class="_word__block_063118"
          >
            <h3
              class="_word__subtitle_063118"
            >
              Meanings
            </h3>
            <ul
              class="_word__definitions_063118"
            >
              <li
                class="_word__definition_063118"
              >
                <p
                  class="_word__definition_063118"
                >
                  News; tidings (used without an article).
                </p>
                <p
                  class="_word__example_063118"
                >
                  Have you had any word from John yet?
                </p>
              </li>
              <li
                class="_word__definition_063118"
              >
                <p
                  class="_word__definition_063118"
                >
                  A promise; an oath or guarantee.
                </p>
                <p
                  class="_word__example_063118"
                >
                  I give you my word that I will be there on time.
                </p>
              </li>
              <li
                class="_word__synonym_063118"
              >
                promise
              </li>
              <li
                class="_word__antonym_063118"
              >
                false promise
              </li>
            </ul>
          </div>
          <div
            class="_word__block_063118"
          >
            <h3
              class="_word__subtitle_063118"
            >
              Synonyms
            </h3>
            <ul
              class="_word__definitions_063118"
            >
              <li
                class="_word__synonym_063118"
              >
                Bible
              </li>
              <li
                class="_word__synonym_063118"
              >
                Logos
              </li>
              <li
                class="_word__synonym_063118"
              >
                vocable
              </li>
            </ul>
          </div>
          <div
            class="_word__block_063118"
          >
            <h3
              class="_word__subtitle_063118"
            >
              Antonyms
            </h3>
            <ul
              class="_word__definitions_063118"
            />
          </div>
          <div
            class="_word__license_063118"
          >
            <p
              class="_word__name_063118"
            >
              CC BY-SA 3.0
            </p>
            <a
              class="_word__url_063118"
              href="https://creativecommons.org/licenses/by-sa/3.0"
            >
              Word License
            </a>
          </div>
          <ul
            class="_word__list_063118"
          >
            <li
              class="_word__item_063118"
            >
              <a
                class="_word__link_063118"
                href="https://en.wiktionary.org/wiki/word"
              >
                Word Source
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
